Add Protobuf chapter (#385)
icerpc/icerpc-docs: content/protobuf/index.md

---
title: Protobuf
description: Using Protocol Buffers with IceRPC
showReadingTime: false
---

## About Protocol Buffers

Protocol Buffers or Protobuf for short is a popular [Interface Definition Language] (IDL) and serialization format

> created by Google and available for many programming languages. [gRPC] and several other RPC frameworks use [Protobuf] as
> their IDL.

Reviewer: ```suggestion
created by Google and available for many programming languages. [Protobuf] is the IDL for [gRPC] and several other RPC frameworks.
```

I think it sounds a bit more natural like this.

This chapter describes the IceRPC + Protobuf integration, not Protobuf itself. The IceRPC + Protobuf integration uses
standard Protobuf files, with the regular Protobuf syntax and semantics.

## The IceRPC + Protobuf integration

The IceRPC + Protobuf integration helps you implement Protobuf services with IceRPC and call such services using (over)
IceRPC.

In practical terms, the IceRPC + Protobuf integration for C#:

- defines an IceRPC-specific client-side and server-side mapping for Protobuf services to C#

> - provides a `protoc` plugin (`protoc-gen-icerpc-csharp`) that generates code for Protobuf services

Author: fixed

- provides a support assembly used by this generated code, `IceRpc.Protobuf.dll`

> - provides a MSBuild integration that helps you execute `protoc` with the `protoc-gen-icerpc-csharp` plugin in MSBuild

Reviewer 2: Because M is pronounced `em`, starting with a vowel sound.
```suggestion
- provides an MSBuild integration that helps you execute `protoc` with the `protoc-gen-icerpc-csharp` plugin in MSBuild
```

projects: `IceRpc.Protobuf.Tools`

User-defined Protobuf types such as messages and enums are handled by the `protoc` compiler and its built-in C# plugin.
The IceRPC + Protobuf integration does not customize this mapping in any way.

The net result is you need to write very little code to call or implement Protobuf services with IceRPC.

For example, let's take a typical Greeter example:

```protobuf
syntax = "proto3";

package visitor_center;
option csharp_namespace = "VisitorCenter";

// A simple Greeter service.
service Greeter {
    // Creates a personalized greeting.
    rpc Greet (GreetRequest) returns (GreetResponse);
}

// The request contains the name of the person to greet.
message GreetRequest {
    string name = 1;
}

// The response contains the greeting.
message GreetResponse {
    string greeting = 1;
}
```

`protoc`'s built-in plugin for C# generates C# classes for the `GreetRequest` and `GreetResponse` messages, while

> `protoc-gen-icerpc-csharp` generates two C# interfaces and a struct for the `Greeter` service:

Reviewer 2: I don't think it's necessary to qualify `classes` and `interfaces` as C# classes and C# interfaces. It's already pretty clear from the context I think.

Or, we should at least be consistent and also qualify 'struct' (in the last line) as `C# struct`

      GreetRequest message,
      IFeatureCollection? features = null,
      CancellationToken cancellationToken = default)
      {
          // implemented by the generated code using IceRPC...
      }
}
```

```csharp {% title="C# generated code - server-side" %}
namespace VisitorCenter;

public partial interface IGreeterService
{
   ValueTask<GreetResponse> GreetAsync(
      GreetRequest message,
      IFeatureCollection features,
      CancellationToken cancellationToken);
}
```

## Next steps

{% grid %}

{% mini-card
   title="Service mapping"
   description="Learn how to call and implement a Protobuf service with IceRPC in C#."
   href="/protobuf/language-mapping/service" /%}

{% mini-card
   title="Protocol Buffers Documentation"
   description="Google's documentation for Protocol Buffers."
   href="https://protobuf.dev/" /%}

{% mini-card
   title="Examples"
   description="See examples of the IceRPC + Protobuf integration for C#."
   href="https://github.com/icerpc/icerpc-csharp/tree/main/examples/protobuf" /%}

{% /grid %}

[gRPC]: https://grpc.io/
[Interface Definition Language]: https://en.wikipedia.org/wiki/Interface_description_language

> [Protobuf]: https://en.wikipedia.org/wiki/Protocol_Buffers

Reviewer 2: Same here, these aren't in the order they're declared.
